documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+    if (isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test B #node", function() {
+    if (!isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test C #live", function() {
+    if (process.env.TEST_MODE) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+});
+```
+
+In case conditionals might appear to be necessary in other scenarios, consider separating the test case into as many test cases as necessary first.
+
+### Using delays
+
+The API methods that we provide should always provide a way for users to know when an operation finishes. For this reason, any method that doesn't immediately respond with a completed operation should use our [Long Running Operations strategy (core-lro)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro).
+
+If any method can't use `core-lro` within reasonable time, and thus we are inclined to use a different delay strategy, we should always wait until the next possible operation is can be fulfilled.
+
+For example, in KeyVault-Keys we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:
+
+```ts
+import { delay } from "@azure/test-utils-recorder";
+// ...
+
+describe("Keys client - restore keys and recover backups", () => {
+  // ...
+  it("can restore a key with a given backup", async function() {
+    // ...
+    await client.createKey(keyName, "RSA");
+    const backup = await client.backupKey(keyName);
+    await client.purgeDeletedKey(keyName);
+    while (true) {
+      try {
+        await client.restoreKeyBackup(backup as Uint8Array);
+        break;
+      } catch (e) {
+        console.log("Can't restore the key since it's not fully deleted:", e.message);
+        console.log("Retrying in one second...");
+        // This delay method comes from the recorder
+        await delay(1000);
+      }
+    }
+    const getResult = await client.getKey(keyName);
+    assert.equal(getResult.name, keyName, "Unexpected key name in result from getKey().");
+    // ...
+  });
+  // ...
+});
+```
+
+Keep in mind that Mocha will have a timeout configuration that will prevent this to run up forever.
+
+The specific `delay` method used in the code above comes from the [Recorder](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder), so that in playback, there will be no delay at all, and tests will pass as soon as possible.
+
+### Exceptions and edge cases
+
+While testing the Azure SDK clients for JavaScript and TypeScript, we should document each client method exceptions through the use of `@throws` in the TypeDoc documentation, and avoid writing test cases for exceptions. Test cases should focus on demonstrating the functionalities of the client, in relation to the functionalities of the service. Exceptions should be in principle documented.
```

Comment:
> avoid writing test cases for exceptions

If we document it (ie, it's part of the customer facing interface) we should have a test asserting that it happens, right? (when possible). Is there a reason we shouldn't write test cases like this?

Reply:
The main issue is that many of the possible service exceptions could be hard to replicate, and we don't want to write mocks representing each one of these cases for now, at least from the feedback I've received. How would we test all of the exceptions? If it's through mocks, do we want to go that way? I'll leave it open to get more feedback, for sure.

Reply:
You've got a good point - I think 'avoid writing test cases for exceptions' should be explained a bit so people can use their discretion. In general though, I agree with you.

The exceptions (minus the message text itself and more any programatically actionable codes) should be tested when we can.

Reply:
I believe that at a minimum, exceptions MUST be tested if they are representative of use cases written by customers. The phrase "exceptions should be tested when we can" does not provide a clear distinction of when this should happen. If we were developing the services, these edge cases would be tested on the services side. We should demand them to provide clear documentation of when exceptions happen. Some of this might currently be available, and based on that we can document how these APIs behave.

Having said that, I do feel odd to avoid writing tests for all of the exceptions, so please consider my response as a request to narrow down what exceptions to test with more clarity.

Reply:
Maybe it makes sense to exercise exceptions we throw in the convenience layer / in helper functions to make sure that guards/checks are working as expected, but not to exercise exceptions that come from the service (e.g. RestError and friends.)

Reply:
That seems good to me! I'll make sure the document reflects that! Thank you, @xirzec